### PR TITLE
Fixes #2508 and #2519: Fixes `uint` scalar handling and overhauls `mod/fmod` for floats

### LINKED ISSUE #2508: incorrect results with `uint` scalars
I discovered this while working #2505, we aren't correctly handing uint scalars. 

Our tests are catching this because I didn't choose the value of the uint scalar very well, so it's being treated as an integer
https://github.com/Bears-R-Us/arkouda/blob/eeebe7095a65b3858b4510c76990101d65b49c8e/tests/operator_tests.py#L31
if we update this to have `"uint64": np.uint64(2**63 + 1)`, we start seeing these incorrect results.

Before updating the test:
```
# ops not implemented by numpy or arkouda: 114
# ops implemented by numpy but not arkouda: 135
# ops implemented by arkouda but not numpy: 4
# ops implemented by both: 611
  Matching results:         611 / 611
  Arkouda execution errors: 0 / 611

  Dtype mismatches:         0 / 611

  Value mismatches:         0 / 611
```

After updating the test:
```
# ops not implemented by numpy or arkouda: 122
# ops implemented by numpy but not arkouda: 147
# ops implemented by arkouda but not numpy: 6
# ops implemented by both: 589
  Matching results:         586 / 589
  Arkouda execution errors: 0 / 589

  Dtype mismatches:         2 / 589
int64(array) + uint64(scalar): float64(np) vs. int64(ak)
int64(array) - uint64(scalar): float64(np) vs. int64(ak)
  Value mismatches:         1 / 589
uint64(scalar) % float64(array): np: [           nan 2.84217094e-14 2.84217094e-14 2.84217094e-14
 2.84217094e-14 6.22222222e-01 2.84217094e-14 7.30158730e-01
 2.84217094e-14 0.00000000e+00]
ak: [nan  0.  0.  0.  0.  0.  0.  0.  0.  0.]
```
So we some combinations involving uint scalars that have appeared to be working, but are actually incorrect. 

I found this by running `ak.arange(10) + np.uint(2**63 + 1)`. We should return a float array to match numpy but we aren't. The updated tests shows this `int64(array) + uint64(scalar): float64(np) vs. int64(ak)`

We also have some operations that we thought supported uint scalars, but don't. We can see that by the number of `ops implemented by numpy but not arkouda` increasing

Fiinally we see there's a value mismatch with `uint scalar % float array`

### LINKED ISSUE #2519: overhaul our `mod` functionality
This has been edited to reflect new requirements for this issues the original is below

Right now our `mod` is incorrect for large scalars `uint/float` and for edge cases in the modulus (`+/-np.inf`, '+/-0.0`)

Attempting to update it to `fmod` correctly handles large scalars but is not incorrect on negative floats and edge cases both being modded and doing the modding.

In this issue we should:
- add `ak.fmod` which matches `np.fmod` using the chpl `fmod` function
- update `ak.mod` to match `np.mod` by doing something similar to `floorDivisionHelper`, by cleverly modifying the return of the chpl `fmod`

ORIGINAL:
While working #2508, I found the section of `test_divmod` that says it's putting the edge cases in the denominator doesn't do that and is identical to the numerator
https://github.com/Bears-R-Us/arkouda/blob/2034dcb323bd731960c9d9b2c40f06fccfc87c5d/tests/stats_test.py#L235-L251

I ran some quick tests on master and `mod` behaves incorrectly with the float edge cases in the denominator
```python
>>> edge_case = [-np.inf, -7.0, np.nan, 7.0, np.inf]
>>> np_edge_case = np.array(edge_case)
>>> ak_edge_case = ak.array(np_edge_case)
>>> ak_ind = ak.arange(1, len(edge_case)+1)
>>> np_ind = np.arange(1, len(edge_case)+1)

# edge cases in numerator (behaves correctly)
>>> ak_edge_case % ak_ind
array([nan 1 nan 3 nan])

>>> np_edge_case % np_ind
array([nan,  1., nan,  3., nan])

# edge cases in denominator (incorrect results)
>>> ak_ind % ak_edge_case
array([nan -5 nan 4 nan])

>>> np_ind % np_edge_case
array([-inf,  -5.,  nan,   4.,   5.])
```

### PR DESCRIPTION
**Pull request by pierce314159**
This PR (fixes #2508 and closes #2519) 
- resolves two bugs with our handling of uint scalars
- overhauls our `mod` and adds `fmod` 
  - adds testing to verify these match their numpy equivalents 
- fixes the `divmod` test

The two bugs involving uint scalars are:
- dtype mismatch:
  - `int arr +/- uint scalar` and vice versa resulted in an int array. It should be a real array to match numpy
  - Resolved 
- value mismatch:
  - `large real scalars (i.e. > 2 ** 63) % real arrays` were giving incorrect values. Since we cast `uint` scalars before doing operations with real arrays, this issue popped up
  - Resolved by shifting results from `fmod`. For more info see:
    - https://github.com/chapel-lang/chapel/issues/22584
    - #2519

Picking more representative examples for uint scalars and uint arrays in `testAllOperators` revealed this incorrect handling.
i.e.
```python
uint scalar: np.uint64(2**63 + 1)
uint array: np.arange(2**64 - SIZE, 2**64, 1, dtype=np.uint64)
```

`run_tests` on master before updating the test:
```
# ops not implemented by numpy or arkouda: 114
# ops implemented by numpy but not arkouda: 135
# ops implemented by arkouda but not numpy: 4
# ops implemented by both: 611
  Matching results:         611 / 611
  Arkouda execution errors: 0 / 611

  Dtype mismatches:         0 / 611

  Value mismatches:         0 / 611
```

`run_tests` on master after updating the test:
```
# ops not implemented by numpy or arkouda: 122
# ops implemented by numpy but not arkouda: 147
# ops implemented by arkouda but not numpy: 6
# ops implemented by both: 589
  Matching results:         584 / 589
  Arkouda execution errors: 0 / 589

  Dtype mismatches:         2 / 589
int64(array) + uint64(scalar): float64(np) vs. int64(ak)
int64(array) - uint64(scalar): float64(np) vs. int64(ak)
  Value mismatches:         3 / 589
uint64(array) % float64(array): np: [           nan 5.68434189e-14 5.68434189e-14 5.68434189e-14
 5.68434189e-14 1.33333333e-01 5.68434189e-14 1.46031746e+00
 5.68434189e-14 0.00000000e+00]
ak: [nan  0.  0.  0.  0.  0.  0.  0.  0.  0.]
uint64(array) % float64(scalar): np: [1.19112358 1.19112358 1.19112358 1.19112358 1.19112358 1.19112358
 1.19112358 1.19112358 1.19112358 1.19112358]
ak: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
uint64(scalar) % float64(array): np: [           nan 2.84217094e-14 2.84217094e-14 2.84217094e-14
 2.84217094e-14 6.22222222e-01 2.84217094e-14 7.30158730e-01
 2.84217094e-14 0.00000000e+00]
ak: [nan  0.  0.  0.  0.  0.  0.  0.  0.  0.]
```

`run_tests` on this PR after updating the tests:
```
# ops not implemented by numpy or arkouda: 122
# ops implemented by numpy but not arkouda: 145
# ops implemented by arkouda but not numpy: 6
# ops implemented by both: 591
  Matching results:         591 / 591
  Arkouda execution errors: 0 / 591

  Dtype mismatches:         0 / 591

  Value mismatches:         0 / 591
```

The biggest difference between the results from master before updating the tests and this PR after updating the tests is `ops implemented by numpy but not arkouda`. We see that this number has increased by 10. I hunted down the ones that it appeared like we supported before. They're all combinations of ops between `uint scalar` and `int array`. For all these operations the `uint array` and `int array` version is also not supported. So I think if we decide to add this support to arkouda we should do it a future issue and add the `vector-vector`, `vector-scalar`, and `scalar-vector` 
```
int64(array) / uint64(scalar)
int64(array) // uint64(scalar)
int64(array) * uint64(scalar)
int64(array) % uint64(scalar)
int64(array) ** uint64(scalar)

uint64(scalar) / int64(array)
uint64(scalar) // int64(array)
uint64(scalar) * int64(array)
uint64(scalar) % int64(array)
uint64(scalar) ** int64(array)
```

### CODE REVIEW

**Review by bmcdonald3 — APPROVED** [review]

File: src/BinOp.chpl [hunk 1]
```diff
@@ -946,29 +964,31 @@ module BinOp
       }
       var repMsg = "created %s".format(st.attrib(rname));
       return new MsgTuple(repMsg, MsgType.NORMAL);
-    } else if (val.type == int && r.etype == uint) {
+    }
+    else if (e.etype == int && val.type == uint) ||
+            (e.etype == uint && val.type == int) {
```
Comment (jaketrookman) [inline src/BinOp.chpl:969]: Can you explain why these cases no longer rely on the `r.etype`.

Comment (pierce314159) [reply src/BinOp.chpl:969]: yah! basically i was updating it to be consistent with the vector scalar case

File: arkouda/numeric.py [hunk 2]
```diff
@@ -584,55 +584,29 @@ def arctan2(num: Union[numeric_scalars, pdarray], denom: Union[numeric_scalars,
...
-                        "scalar": num,
-                        "denom": denom,
-                        "dtype": resolve_scalar_dtype(num),
-                    },
-                ),
-            )
+    if not all(isSupportedNumber(arg) or isinstance(arg, pdarray) for arg in [num, denom]):
+        raise TypeError(
+            f"Unsupported types {type(num)} and/or {type(denom)}. Supported "
+            "types are numeric scalars and pdarrays. At least one argument must be a pdarray."
```
Comment (Ethan-DeBandi99) [inline arkouda/numeric.py:590]: I would expect this to be handled by the type checking. Is there something causing it to not handle this as expected?

File: arkouda/pdarrayclass.py [hunk 3]
```diff
@@ -3075,6 +3078,75 @@ def sqrt(pda: pdarray, where: Union[bool, pdarray] = True) -> pdarray:
...
+    divisor : numeric scalars or pdarray
+        The array that will be the bases for the modular division.
+
+    Returns
+    -------
+    pdarray
+        Returns an array that contains the element-wise remainder of division.
+    """
+    if not builtins.all(
+        isSupportedNumber(arg) or isinstance(arg, pdarray) for arg in [dividend, divisor]
```
Comment (Ethan-DeBandi99) [inline arkouda/pdarrayclass.py:3124]: Same question here. I would expect the `@typechecked` to handle this.

**Review by Ethan-DeBandi99 — COMMENTED** [review]
Just one question to clarify. Otherwise, everything looks good.

**Review by jaketrookman — APPROVED** [review]
Looks good. Great work

File: arkouda/numeric.py (hunk 2 above)
Comment (pierce314159) [reply arkouda/numeric.py:590]: yeah i was surprised by this too, but the inclusion of `numeric_scalars` in the type hint make it so it doesn't do any type checking (at least from what I saw). I was thinking we would want to take a look at similar places where we do type hints with groups that aren't classes and see how we want to handle them. I can make an issue to dig into that if u agree with that way forward

Comment (Ethan-DeBandi99) [reply arkouda/numeric.py:590]: I think investigating this under a separate issue makes sense.

Comment (pierce314159) [reply arkouda/numeric.py:590]: Captured in #2528

File: arkouda/pdarrayclass.py (hunk 3 above)
Comment (pierce314159) [reply arkouda/pdarrayclass.py:3124]: Captured in #2528

**Review by Ethan-DeBandi99 — APPROVED** [review]
Looks good. The one component that we will be researching further has been captured in an to look into.
